### PR TITLE
Implement a rate limiter to the dbus calls

### PR DESCRIPTION
**Pull request by Philip-Scott**
## Background info:

In Lights-Up, i'm using the philips hue lights and it's API is limited to only one call per 200ms. The problem is that just connecting an API call to a slider's value changed signal causes a call for even the slightest change. This usually isn't a problem in normal apps, but when talking to an external API you need to limit said requests in some form or another. 

https://github.com/Philip-Scott/Lights-Up/blob/master/src/api/Login.vala

## What this PR does:

Currently, if you start to slowly move the temperature slider, you don't really get any feedback of what the current value is until you're done changing the value. This PR makes it so that subsequent calls are called 300ms apart, this allowing for the color of the screen to start to slowly change even while you move the slider. 

RateLimiter allows the first call to go though, and will only allow the last call made with the same ID within the timeout window to go though.  The class might be more complicated than what NightLight needs, but it's because it was made with multi-threading and support for keeping track of multiple types of calls in mind. If we want this behavior i can simplify the class here :) 

## How to test
See the difference between moving slowly the slider on the Switchboard plug, and the one here in this PR :)

### CODE REVIEW

File: src/Services/RateLimiter.vala [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+ * License along with this program; if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA
+ */
+
+public class Nightlight.RateLimiter : Object {
+    /**
+     * Time in ms between the execution of events
+     */
+    public int timeout { get; set; default = 300; }
```
Comment (donadigo) [inline src/Services/RateLimiter.vala]: Any reason for this property to be public? Could we just make this a constant?

File: src/Services/RateLimiter.vala [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+    private Queue<QueuedEvent> event_queue;
+    private HashTable<string, QueuedEvent> event_set;
+    private uint? timeout_id = null;
+
+    private Mutex mutex = Mutex ();
+
+    construct {
+        event_queue = new Queue<QueuedEvent>();
```
Comment (donadigo) [inline src/Services/RateLimiter.vala]: Spaces between `()`.

File: src/Services/RateLimiter.vala [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        event_set = new HashTable<string, QueuedEvent>(str_hash, str_equal);
```
Comment (donadigo) [inline src/Services/RateLimiter.vala]: Same here.

File: src/Services/RateLimiter.vala [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+        if (event != null) {
+            event.run ();
+        }
+
+        timeout_id = Timeout.add (timeout, () => {
+            timeout_id = null;
+
+            mutex.lock ();
+            var length = event_queue.length;
```
Comment (donadigo) [inline src/Services/RateLimiter.vala]: Why is this enclosed with mutex locking?

**Review by donadigo — DISMISSED** [review]
Look diff comments. In addition to that, you should use libgee instead of GLib collections. Otherwise, works well.

**Comment by jeremypw** [comment]
The RateLimiter class seems to be overkill here - moreover it would have to be replicated in the Switchboard plug, which has the same temperature control functionality.  I have replaced it with a simple timeout and could not see any difference in the effect.
I think the gsettings backend has some degree of throttling anyway but maybe not sufficient for this purpose.

**Comment by jeremypw** [comment]
As I contributed to the code, this needs a different reviewer to approve (or not).

**Review by danrabbit — APPROVED** [review]
I don't see a huge difference here personally, but couldn't hurt to limit I suppose

**Pull request merged**